crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,861 @@
+ */
+static int find_server_cert(X509_STORE *ts, /* may be NULL */
+                            STACK_OF(X509) *untrusted, /* may be NULL */
+                            const OSSL_CMP_MSG *msg,
+                            STACK_OF(X509) *found_certs)
+{
+    STACK_OF(X509) *trusted;
+
+    if (!ossl_assert(msg != NULL && found_certs != NULL))
+        return 0;
+
+    if (!find_acceptable_certs(untrusted, msg, ts, found_certs))
+        return 0;
+
+    if (ts != NULL) {
+        int ret;
+
+        trusted = ossl_cmp_X509_STORE_get1_certs(ts);
+        ret = find_acceptable_certs(trusted, msg, ts, found_certs);
+        sk_X509_pop_free(trusted, X509_free);
+        if (ret == 0)
+            return 0;
+    }
+
+    ossl_cmp_add_error_line(sk_X509_num(found_certs) ?
+                            "found at least one matching server cert" :
+                            "no current matching server cert found");
+    return 1;
+}
+
+/*
+ * Exceptional handling for 3GPP TS 33.310 [3G/LTE Network Domain Security
+ * (NDS);Authentication Framework (AF)], only to use for IP and if the ctx
+ * option is explicitly set: use self-signed certificates from extraCerts as
+ * trust anchor to validate server cert - provided it also can validate the
+ * newly enrolled certificate
+ */
+static int srv_cert_valid_3gpp(OSSL_CMP_CTX *ctx, X509 *scrt,
+                               const OSSL_CMP_MSG *msg)
+{
+    int valid = 0;
+    X509_STORE *store = X509_STORE_new();
+
+    if (store != NULL /* store does not include CRLs */
+            && ossl_cmp_X509_STORE_add1_certs(store, msg->extraCerts,
+                                              1 /* self-signed only */))
+        valid = OSSL_CMP_validate_cert_path(ctx, store, scrt);
+
+    if (valid) {
+        /*
+         * verify that the newly enrolled certificate (which is assumed to have
+         * rid == 0) can also be validated with the same trusted store
+         */
+        EVP_PKEY *privkey = OSSL_CMP_CTX_get0_newPkey(ctx, 1);
+        OSSL_CMP_CERTRESPONSE *crep =
+            ossl_cmp_certrepmessage_get0_certresponse(msg->body->value.ip, 0);
+        X509 *newcrt = ossl_cmp_certresponse_get1_certificate(privkey, crep);
+        /*
+         * maybe better use get_cert_status() from cmp_client.c, which catches
+         * errors
+         */
+        valid = OSSL_CMP_validate_cert_path(ctx, store, newcrt);
+        X509_free(newcrt);
+    }
+    X509_STORE_free(store);
+    return valid;
+}
+
+/* Find and validate any potentially usable server cert */
+static X509 *find_srvcert(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+{
+    X509 *scrt = NULL;
+    int valid = 0;
+    GENERAL_NAME *sender = msg->header->sender;
+
+    if (sender == NULL || msg->body == NULL)
+        return 0; /* other NULL cases already have been checked */
+    if (sender->type != GEN_DIRNAME) {
+        CMPerr(0, CMP_R_SENDER_GENERALNAME_TYPE_NOT_SUPPORTED);
+        return NULL; /* FR#42: support for more than X509_NAME */
```

Comment:
What does "FR#42" mean?

Reply:
This is a rusted reference to a feature request: https://github.com/mpeylo/cmpossl/issues/96

Reply:
Confusing comment removed.